content/3_Silver/Func_Graphs.mdx

```diff
@@ -444,51 +445,181 @@ Try simulating the algorithm on the following directed graph in CSAcademy's
+		 */
+		if (ans[child] == -1)
+		{
+			ans[child] = ans[x];
+			fill_radj(child);
+		}
 	}
 }
 
-void go(int x) {
+void floyd(int x)
+{
 	int y = x;
-	do { // find a cycle using Floyd's algo
-		x = P[x];
-		y = P[P[y]];
+	do
+	{  // find a cycle using Floyd's algorithm
+		x = adj[x];
+		y = adj[adj[y]];
 	} while (y != x);
-	do { // set ans[x] = x for all x along cycle
+	do
+	{  // set ans[x] = x for all x along cycle
 		ans[x] = x;
-		x = P[x];
+		x = adj[x];
 	} while (y != x);
-	do { // set ans'es for all x not along cycle
+	do
+	{  // set ans'es for all x not along cycle
 		fill_radj(x);
-		x = P[x];
+		x = adj[x];
 	} while (y != x);
 }
 
-int main() {
-	setIO(); rv(N,P); trav(t,P) --t;
-	radj.rsz(N); F0R(i,N) radj[P[i]].pb(i);
-	ans = vi(N,-1); F0R(i,N) if (ans[i] == -1) go(i);
-	trav(t,ans) pr(t+1,' ');
+int main()
+{
+	cin.tie(0)->sync_with_stdio(false);
+	cin >> N;
+
+	adj.assign(N, -1);
+	for (auto &e : adj)
+	{
+		cin >> e;
+		e--;
+	}
+
+	/* 
+	 * For each node, we have to use a vector to store its children;  
+	 * at nodes combining the cycle with other parts of the connected component,
+	 * there would be more than one outgoing arrow in the reversed adjacency list
+	 */
+	radj.assign(N, {});
+	for (int i = 0; i < N; i++)
+		radj[adj[i]].push_back(i);
+
+	ans.assign(N, -1);
+	// we run Floyd's algorithm for each connected component
+	for (int i = 0; i < N; i++)
+		if (ans[i] == -1)
+			floyd(i);
+
+	for (auto &a : ans)
+		cout << a + 1 << " ";
 }
 ```
 
 </CPPSection>
 
+<JavaSection>
+
+```java
+import java.util.*;
+import java.io.*;
+
+public class Badge {
+	static Integer[] adj, ans;
+	/*
+	 * For each node, we need a list to store its children; at nodes
+	 * combining the cycle with other part of the connected component, there would
+	 * be more than one outgoing arrow in the reversed adjacency list
+	 */
+	static List<List<Integer>> radj;
+
+	public static void main(String[] args) throws IOException {
+		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
+
+		int N = Integer.parseInt(in.readLine());
+		adj = new Integer[N];
+		ans = new Integer[N];
+		radj = new ArrayList<>();
+
+		StringTokenizer st = new StringTokenizer(in.readLine());
+		for (int i = 0; i < N; i++) {
+			adj[i] = Integer.valueOf(st.nextToken()) - 1;
+			ans[i] = -1;
+			radj.add(new ArrayList<>());
+		}
+
+		for (int i = 0; i < N; i++)
+			radj.get(adj[i]).add(Integer.valueOf(i));
+
+		for (int i = 0; i < N; i++) {
+			// run Floyd's algorithm on every connected component
+			if (ans[i] == -1)
+				floyd(i);
+		}
+
+		for (Integer i : ans) {
+			System.out.print(i + 1);
+			System.out.print(" ");
+		}
+	}
+
+	private static void floyd(int x) {
+		int a = adj[x];
+		int b = adj[adj[x]];
+
+		// find a cycle using Floyd's algorithm
+		while (a != b) {
+			a = adj[a];
+			b = adj[adj[b]];
+		}
+
+		// for each node a in the cycle, the answer ans[a] will be a as well
+		do {
```

Comment:
does it have to be do while? I've never seen a situation where you have to use one

Reply:
probably can, but it doesn't really matter(?)

Reply:
not really, but then you will have 2 more lines: 
```java
ans[a] = a;
a = adj[a];
while (a != b) {
	ans[a] = a;
	a = adj[a];
} 
```